### DISCUSSION TITLE
Is there an equivalent of WPF's CompositionTarget.Rendering event in AvaloniaUI?

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#13697] [Q&A] [answered]
Hi, as stated in the title, is there an equivalent of WPF's CompositionTarget.Rendering event in AvaloniaUI?

I used to use it to check some states only once per frame, such as:
```
private void OnCompositionTargetRendering(object? sender, EventArgs e)
{
    if (NeedsRefresh)
    {
        NeedsRefresh = false;
        UpdateImageAsync();
    }
}
```
Here NeedsRefresh may have been set to true severals times in the same frame.
Previously I was using a timer to "group" changes in a small interval range to avoid to update too often, but Composition.Rendering is more reactive than a timer interval...

**Comment by stevemonaco** [comment, marked as answer]
`CompositionTarget.Rendering` doesn't exist, but you can use `TopLevel.RequestAnimationFrame` to tie into the compositor frame rate.

```cs
public MainWindow()
{
    InitializeComponent();
    RequestAnimationFrame(Tick);
}

public void Tick(TimeSpan time) // total elapsed time, not frame time
{
    textBlock.Text = time.TotalMilliseconds.ToString();
    RequestAnimationFrame(Tick);
}
```

**Reply by dbriard** [reply]
Look nice, thanks!
